sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -125,6 +125,17 @@ public class ServiceBusProcessor : IAsyncDisposable
         /// successful execution of the operation; otherwise, false.</value>
         public virtual bool AutoCompleteMessages { get; }
 
+        /// <summary>
+        /// Gets or sets whether the messages bodies use pooled byte arrays underneath in case of binary data.
+        /// When this option is enabled <see cref="ServiceBusReceivedMessage"/> cannot be safely kept for longer than the
```

Comment:
Is it that the message body cannot be held onto rather than the entire message?

Reply:
I tried to formulate it from the perspective of the user and the most likely use case. Actually you cannot hold the body of the annotated message but since that is "owned" by the `ServiceBusReceivedMessage` I felt this would be too much unnecessary details that doesn't help. What is directly exposed to the user is the received message and like you pointed out someone might try to hold it longer than the lifetime of the event handler delegate. 

Of course if we want to be precise as possible one could argue that anything that is not directly related to binary data inside the body can be kept. Should we go to that level of detail in the description?

Reply:
I disagree about the received message owning the underlying AMQP message.   The AMQP message needs to be able to stand-alone and outlive any higher level type that contains it.  One of the goals that we're aiming for is allowing developers to pick up and move the AMQP message between the messaging brokers and/or between namespaces.   For example, something like:

```csharp
var receivedMessage = someServiceBusReceiver.ReceiveMessageAsync();
var newMessage = ServiceBusMessage.FromRawAmqpMessage(receivedMessage.GetRawAmqpMessage());

await otherNamespaceSender.SendMessageAsync(newMessage));
```

Or:

```csharp
var receivedMessage = someServiceBusReceiver.ReceiveMessageAsync();
var eventData = EventData.FromRawAmqpMessage(receivedMessage.GetRawAmqpMessage());

await eventHubProducer.SendAsync(new[] { eventData });;
```

Granted, we could do a full copy, but in my opinion part of the value of a neutral shared representation of the protocol message is the ability to share between libraries that use that protocol.

Reply:
Any suggestions how to better convey what happens when this flag is enabled without mentioning the ownership? The original comment from @JoshLove-msft was

> but what if a user holds on to the message reference after their callback returns? The body would be gone.

which lead me down to formulate it the way it is currently formulated. By the way the scenarios you showed are still possible even when the flag is enabled. But they are only possible within the scope of the handler

Reply:
> But they are only possible within the scope of the handler

That's the case that concerns me.  Take my simple example and expand it out to something more real-world:

```csharp
public class SomeClass
{
    EventHubDataBatch _activeBatch;

    private async Task MessageHandler(ProcessMessageEventArgs args)
    {
         activeBatch ??= (await producer.CreateBatchAsync());

         if (!_activeBatch.TryAdd(EventData.FromRawAmqpMessage(args.Message.GetRawAmqpMessage()))
         {
             await producer.SendAsync(_activeBatch);
             _activeBatch.Dispose();
             _activeBatch = null;
         }
    }
}
```

Wouldn't boxing the body and holding it in a class member prevent disposal?

Reply:
The case you show could be easily addressed by making sure the appropriate conversion methods take take this case into account and then copy the payload from the buffer. So for example with ServiceBus a concrete example could be creating a `ServiceBusMessage` from a `ServiceBusReceivedMessage`

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusMessage.cs#L63-L72

The `MessageBody` methods could detect the pooling body type and then copy the payload over. Then you can keep it around as long as you want. 

More problematic would be this case

```
public class SomeClass
{
    ConcurrentQueue< AmqpAnnotatedMessage> amqpMessages;

    private async Task ServiceBusProcessorMessageHandler(ProcessMessageEventArgs args)
    {
         amqpMessages??= new ConcurrentQueue<AmqpAnnotatedMessage>();
         amqpMessages.Enqueue(args.Message.GetRawAmqpMessage());
         ...
    }
}
```

because there would be no way to protect against this scenario. Yet I'm wondering why should you? It boils down to the option is entirely opt-in and can be controlled on a case by case basis. So if someone opts-in for the high throughput/low allocation behavior they have to follow certain rules and I believe they are not too invasive. 

The concept of scopes is well known in .NET. IoC containers have for example request scopes. ASP.NET Core has scoped objects with a certain lifetime too and if users throw the request into a state capturing entity that lives longer than the controler/request things will not play nice ;) Another example is the HttpContextAccessor and it's guidelines

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-5.0#httpcontext-access-from-a-background-thread

To make users aware that opted in for this behavior and might still try to abuse things it would be possible to kind of "arm" the internal body after release to throw a meaningful exception. Then these types of super duper edge cases would be found early in DEVTEST or QA environments.

Reply:
I would agree that the opt-in nature of this feature, makes it reasonable that a message cannot be accessed outside of the handler. In practice, I would expect the vast majority of users to be doing this anyway independent of this new flag. But the new flag would surely be necessary to introduce a semantic change like this. We will need to get architect sign off on this feature of course.

Reply:
This is something that trips me up with English always:

>  I would expect the vast majority of users to be doing this anyway independent of this new flag

Does it mean you expect the majority of the users accessing the message outside the handler or you don't expect the majority of the users to access the message outside the handler.

Reply:
> Yet I'm wondering why should you? It boils down to the option is entirely opt-in and can be controlled on a case by case basis. 

It's a fair enough point.  The trick comes back to naming and documenting this to make sure that it gets used in the common case.  I feel like we're in the `ConfigureAwait` scenario, where we would want this to be used more often than not, but have to disable by default to guard against the uncommon, but less obvious, case.

Reply:
> This is something that trips me up with English always:
> 
> > I would expect the vast majority of users to be doing this anyway independent of this new flag
> 
> Does it mean you expect the majority of the users accessing the message outside the handler or you don't expect the majority of the users to access the message outside the handler.

Nope, my wording was imprecise. I mean that I expect most users to only interact with the message inside the handler.

Reply:
Had some thoughts recently. What if we would manage the buffer associated with a message by using a conditional weak table on the receiver and clean the buffers behind the scenes ones the received messages is gone (GCed)?